### PR TITLE
Add transpose conversions for nested Option and Result

### PR DESCRIPTION
**Pull request by cramertj**
These impls are useful when working with combinator
methods that expect an option or a result, but you
have a `Result<Option<T>, E>` instead of an `Option<Result<T, E>>`
or vice versa.

### CODE REVIEW

**Comment by cuviper** [comment]
The general idea came up in the users forum too, including a few ways to generically convert `Try` types: https://users.rust-lang.org/t/convenience-method-for-flipping-option-result-to-result-option/13695

**Comment by TimNN** [comment]
The implementation looks good to me (although the last test is a bit hard to follow).

cc @rust-lang/libs regarding if we want these impls or not. (Does this need a crater run?)

**Comment by alexcrichton** [comment]
The @rust-lang/libs team discussed this during triage today and our conclusion was that we may not wish for these methods in precisely this form. We were having difficulty (but @cramertj feel free to fill in the gaps) coming up with situations where you'd use this via something like `.into()`. We did think, however, that these could be useful as inherent methods on Option/Result types though, so maybe they'd be better there rather than via the `From` trait?

**Comment by dtolnay** [comment]
We were having difficulty coming up with situations where you would need these as `Into` conversions in a generic context.

```rust
// Nobody would write this.
fn f<T, E, R: Into<Result<Option<T>, E>>>(input: R) {
```

If we don't need them to be generic this way then we don't need them to be trait impls, so inherent methods could be a more discoverable solution and also behave better where type inference around an into-conversion would be ambiguous.

```rust
impl<T, E> Result<Option<T>, E> {
    pub fn transpose(self) -> Option<Result<T, E>>;
}

impl<T, E> Option<Result<T, E>> {
    pub fn transpose(self) -> Result<Option<T>, E>;
}
```

**Comment by cramertj** [comment]
Yeah I think that makes sense! `transpose` or something seems good to me.

**Comment by shepmaster** [comment]
People have certainly asked for the general functionality:

- [Is it possible to convert Option<Result<T, E>> to a Result<Option<T>, E> without using match?](https://stackoverflow.com/q/43166393/155423)

I know there's one for `Result<Option<T>, E>` somewhere, but not able to find it at the moment, so I'm in favor of some type of functionality.

**Comment by shepmaster** [comment]
It sounds like the team is OK with the inherent methods, so I've updated the labels to reflect that. Please change it back if I read incorrectly.

**Comment by kennytm** [comment]
Review ping for you @TimNN!

File: src/test/run-pass/result-opt-conversions.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    y = None;
+    assert_eq!(x, y.transpose());
+    assert_eq!(x.transpose(), y);
+
+    x = Err(BadNumErr);
+    y = Some(Err(BadNumErr));
+    assert_eq!(x, y.transpose());
+    assert_eq!(x.transpose(), y);
+
+    let res: Result<Vec<i32>, BadNumErr> =
```
Comment (TimNN) [inline src/test/run-pass/result-opt-conversions.rs]: I find this part of the to be extremely hard to follow. Is there a specific purpose of this test? Is this a (simplified) real-world use case? Or is that supposed to test type inference?

Comment (cramertj) [reply src/test/run-pass/result-opt-conversions.rs]: This is a simplified version of a real use-case I had. I arrived at this solution when trying to handle possible errors in a `filter_map` closure.

**Comment by golddranks** [comment]
Just for the record, there was an earlier issue about a `try_map` and `flip` methods: https://github.com/rust-lang/rfcs/issues/1815 It lead to this crate that provides similar functionality than this conversion: https://crates.io/crates/try_map It's great to see that this is going to be in `std`, since this is a commonly needed functionality, but there's always a threshold to find & use external crates for trivial things.
